File: leetcode/questions/200/java/200.java
import java.util.ArrayDeque;

class Solution {
    private void markIsland(char[][] grid, int row, int col) {
        int rows = grid.length, cols = grid[0].length;
        int[][] steps = {{0, 1}, {0, -1}, {-1, 0}, {1, 0}};
        ArrayDeque<int[]> stack = new ArrayDeque<>();
        stack.add(new int[]{row, col});

        while(!stack.isEmpty()) {
            int[] pair = stack.pollLast();
            int r = pair[0], c = pair[1];

            if (r < 0 || r >= rows || c < 0 || c >= cols || grid[r][c] != '1') {
                continue;
            }

            grid[r][c] = 'x';

            for (int[] step: steps) {
                int newRow = r + step[0], newCol = c + step[1];

                if (newRow >= 0 && newRow < rows && newCol >= 0 && newCol < cols && grid[newRow][newCol] == '1') {
                    stack.addFirst(new int[]{newRow, newCol});
                }
            }
        }
    }

    public int numIslands(char[][] grid) {
        int cntIslands = 0;

        for (int row = 0; row < grid.length; ++row) {
            for (int col = 0; col < grid[0].length; ++col) {
                if ( grid[row][col] == '1') {
                    cntIslands += 1;
                    markIsland(grid, row, col);
                }
            }
        }

        return cntIslands;
    }
}

File: leetcode/questions/200/python/200.py
from typing import List

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        rows, cols = len(grid), len(grid[0])
        steps = ((0, 1), (0, -1), (1, 0), (-1, 0))

        def mark_island(row, col):
            stack = [(row, col)]

            while stack:
                r, c = stack.pop()

                if not (0 <= r < rows and 0 <= c < cols and grid[r][c] == "1"):
                    continue

                grid[r][c] = "-1"

                for r_inc, c_inc in steps:
                    new_r, new_c = r + r_inc, c + c_inc

                    if 0 <= new_r < rows and 0 <= new_c < cols and grid[new_r][new_c] == "1":
                        stack.append((new_r, new_c))

        cnt_islands = 0

        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == "1":
                    mark_island(r, c)
                    cnt_islands += 1

        return cnt_islands